### ISSUE TITLE
Searching for dupes, then searching again, changes the data shown and is confusing (with caching on)

### ISSUE THREAD

**Issue by alexhunsley** [qarmin/czkawka#1085]
**Desktop**
 - Czkawka version: 6.0.0 gui
 - OS version: Mac OS 13.5.2
 - hardware: MacBook Pro 2021 (apple silicon)
 - compile and run method: as per instructions, having previously done `brew install rustup`

**Bug Description**

Ensure caching is on.

Create a folder "reference" with some test data (images or whatever), with no dupes in it. Copy the folder to another called "other".

Add the two folders into czkawka. Select "Reference" checkbox next to the "reference" folder.

Do a duplicate file search and then "Select -> Select all". The results look sensible: there's a ref folder listed above every dupe in a non-reference folder.

<img width="1098" alt="Screenshot 2023-10-12 at 12 31 43" src="https://github.com/qarmin/czkawka/assets/793071/e0134fb5-75a6-4b9c-8f09-ada21862d220">

Now click search again, and the display changes: there is only one reference folder file listed, and a bunch of matches in the other folder. Do "Select -> Select all". It will select everything except the one reference folder file.

<img width="1102" alt="Screenshot 2023-10-12 at 12 30 51" src="https://github.com/qarmin/czkawka/assets/793071/26c9490b-d5f3-403d-a520-b5c79826e115">

This is confusing; it now looks to me that this means hitting 'delete' will delete original data, because the 'other' folder matches are non-matching files amongst themselves, which are all now ticked. In fact, I don't think original data *is* going to be deleted; it's just that some dupes in the reference folder are now missing from the display.

Basically, hitting search should be idempotent: the result shouldn't change if you haven't changed anything since the last search.

When I turn caching off and delete existing cache files, I don't get this issue.

P.S. another issue, which I haven't go repro steps for but may be related to the issue above: I've search for dupes than deleted them, then searched again immediately with all same settings and more duplicates come up. I've seen this 2-3 times now.

**Comment by alexhunsley** [comment]
Can the cli version do "reference folders" btw, or is that just in the gui?

**Comment by qarmin** [comment]
Reference folders is a core feature, so is theoretically available both for gui and cli, but for now only gui have interface to use it.

**Comment by alexhunsley** [comment]
Are there unit tests for depuplicating functionality btw? Had a quick look and couldn't see anything

**Comment by qarmin** [comment]
For now tests are quite basic:
- https://github.com/qarmin/czkawka/blob/9b57382e39d81eabc934fb23aebd26908fb47af8/.github/workflows/linux_cli.yml#L43-L49
- https://github.com/qarmin/czkawka/blob/master/czkawka_core/src/duplicate.rs#L1236
but slowly starts to be more advanced (e.g. with https://github.com/qarmin/czkawka/pull/1083)

**Comment by qarmin** [comment]
Which exactly version did you checked?
Official 6.0.0 or custom build?

I found problem with latest master version(unreleased 6.1.0 version), but I remember that I already touched this code, so it could be recent regression.

**Comment by alexhunsley** [comment]
Official 6.0.0

**Comment by alexhunsley** [comment]
Btw, I might use blackbox-command-tester to do some tests for czkawka.

https://github.com/alexhunsley/blackbox-command-tester

If you use "record" mode it acts as a snapshot tester, too.

**Comment by qarmin** [comment]
Current regression tester in CI already have similar functionality, but done in different way.

If finds duplicates, then remove them in specific order and at the end compare list of files to check if proper files were deleted.

CLI output is not stable, especially with bigger number of groups may randomly reorder elements, depending on order of checked items.

**Comment by alexhunsley** [comment]
Ah sorry, my previous comment was wrong; I'd checked out and built the current main branch about 4-5 days ago, so yes it might be showing the recent bug you mentioned!

**Comment by alexhunsley** [comment]
Ok, I've runt he GUI compiled from the exact 6.0.0 tag and I can't reproduce the problem described in this ticket (when I have caching enabled). So yeah it looks like an issue added since 6.0.0 tag.
